### PR TITLE
Implement vi mode jumps

### PR DESCRIPTION
**Pull request by Yaraslaut**
Closes https://github.com/contour-terminal/contour/issues/1101

### CODE REVIEW

File: src/vtbackend/ViInputHandler.h [hunk 1]
```diff
@@ -100,6 +100,9 @@ enum class ViMotion : uint8_t
     ToCharLeft,            // F {char}
     RepeatCharMove,        // ;
     RepeatCharMoveReverse, // ,
+    JumpToLastJumpPoint,   // '' or `` (jump to last jump)
+    JumpToMarkUp,          // <C-O>
+    JumpToMarkDown,        // <C-I>
```
Comment (christianparpart) [inline src/vtbackend/ViInputHandler.h:105]: I think that Up/Down is really misleading here. It should be either Out/In or Backward/Forward, which makes more sense in my opinion :)

File: metainfo.xml [hunk 2]
```diff
@@ -113,6 +113,7 @@
           <li>Protect user from accidentally pasting too large input (#1198)</li>
           <li>Add binding to exit normal mode with `Esc` (#1604)</li>
           <li>Add config option to switch into insert mode after yank (#1604)</li>
+          <li>Add vi-like normal mode Jumps (#1101)</li>
```
Comment (christianparpart) [inline metainfo.xml:116]: we should probably mention what that means...

 `('', <C-O>, <C-I>`)

File: src/vtbackend/JumpHistory.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
+#include <vtbackend/JumpHistory.h>
```
Comment (christianparpart) [inline src/vtbackend/JumpHistory.cpp:1]: license header missing

File: src/vtbackend/JumpHistory.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
+#pragma once
```
Comment (christianparpart) [inline src/vtbackend/JumpHistory.h:1]: license header missing

File: src/vtbackend/JumpHistory.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+#include <vtbackend/primitives.h>
+
+#include <vector>
+
+namespace vtbackend
+{
+
+struct JumpHistory
```
Comment (christianparpart) [inline src/vtbackend/JumpHistory.h:10]: if we have private data/methods, it should be a `class` instead (ref: CppCoreGuidelines IIRC)

File: src/vtbackend/JumpHistory.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+{
+
+    template <typename T>
+    void add(T&& cell)
+    {
+        applyOffset();
+        _history.push_back(std::forward<decltype(cell)>(cell));
```
Comment (christianparpart) [inline src/vtbackend/JumpHistory.h:17]:
```suggestion
        _history.push_back(std::forward<T>(cell));
```

File: src/vtbackend/ViCommands.cpp [hunk 7]
```diff
@@ -773,8 +782,13 @@ CellLocation ViCommands::globalCharDown(CellLocation location, char ch, unsigned
     return result;
 }
 
-CellLocation ViCommands::translateToCellLocation(ViMotion motion, unsigned count) const noexcept
+CellLocation ViCommands::translateToCellLocation(ViMotion motion, unsigned count) noexcept
```
Comment (christianparpart) [inline src/vtbackend/ViCommands.cpp:785]: requiring both functions to be mutable sounds like a bad design choice. they were meant to *only* translate. But I do think I know why you where doing it, as these functions translate AND record cell location.

So either we factor out the recording of the current cell location, or (IMHO), I'd suggest to rename those function to something e.g. like `translateToCellLocationAndRecord`. What do you think?

**Review by christianparpart — COMMENTED** [review]
Thank you very much, @Yaraslaut. I've tested it locally, and it's amazing. I only left very few comments though :)

File: src/vtbackend/ViCommands.cpp (hunk 7 above)
Comment (Yaraslaut) [reply src/vtbackend/ViCommands.cpp:785]: renamed to `translateToCellLocationAndRecord`

**Review by christianparpart — APPROVED** [review]

**Pull request merged**